File: lib/src/engine.dart
import 'package:teno_textsearch/src/sinks.dart';

import 'extension.dart';
import 'index_map.dart';
import 'tokenizer.dart';

class SearchResult {
  final String key;
  final String field;
  const SearchResult(this.key, this.field);

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    return (other is SearchResult) && other.key == key && other.field == field;
  }

  @override
  int get hashCode => Object.hash(key, field);

  @override
  String toString() {
    return 'SearchResult(key: $key, field: $field)';
  }
}

class TenoTextSearch {
  late Tokenizer _tokenizer;
  late IndexMap _indexMap;
  final String indexKey;
  final List<String> indexFields;
  final bool caseSensitive;
  final bool forwardedOnly;

  TenoTextSearch({
    int minChars = 3,
    int maxChars = 32,
    String wordDelimiter = ' ',
    this.caseSensitive = false,
    this.forwardedOnly = true,
    required this.indexKey,
    required this.indexFields,
  }) {
    _tokenizer = Tokenizer(
        minChars: minChars,
        maxChars: maxChars,
        delimiter: wordDelimiter,
        forwardedOnly: forwardedOnly,
        caseSensitive: caseSensitive);
    _indexMap = IndexMap(tokenizer: _tokenizer, fields: indexFields);
  }

  void index(Map<String, dynamic> json) {
    final key = json[indexKey];
    assert(key?.isNotEmpty == true, 'Index key must exist in json');
    final mappedJson = json.pick(indexFields);
    if (mappedJson.isNotEmpty) {
      _indexMap.add(key, mappedJson);
    }
  }

  Iterable<SearchResult> search(String text) {
    final words = text.split(_tokenizer.delimiter);
    final result = words.flatMap((element) =>
        _indexMap.find(caseSensitive ? element : element.toLowerCase()) ??
        <List<int>>[]);
    return result.map(
        (e) => SearchResult(_indexMap.keyOf(e[0]), _indexMap.fieldOf(e[1])));
  }

  exportIndexToFile(String path) {
    final fileSink = FileSink(path);
    _indexMap.export(fileSink);
  }

  importFromJson(Map<String, dynamic> json) {
    _indexMap.import(json);
  }
}

File: js/src/engine.ts
import Tokenizer from './tokenizer'; 
import IndexMap, { DocPropIndex } from './index_map';

class SearchResult {
  readonly key: string;
  readonly field: string;

  constructor(key: string, field: string) {
    this.key = key;
    this.field = field;
  }

  equals(other: unknown): boolean {
    return other instanceof SearchResult && other.key === this.key && other.field === this.field;
  }

  toString(): string {
    return `SearchResult(key: ${this.key}, field: ${this.field})`;
  }
}

class TenoTextSearch {
  private _tokenizer: Tokenizer;
  private _indexMap: IndexMap;
  readonly indexKey: string;
  readonly indexFields: string[];
  readonly caseSensitive: boolean;

  constructor({
    minChars = 3,
    maxChars = 32,
    wordDelimiter = ' ',
    caseSensitive = false,
    indexKey,
    indexFields,
  }: {
    minChars?: number;
    maxChars?: number;
    wordDelimiter?: string;
    caseSensitive?: boolean;
    indexKey: string;
    indexFields: string[];
  }) {
    this._tokenizer = new Tokenizer({
      minChars,
      maxChars,
      delimiter: wordDelimiter,
      caseSensitive,
    });
    this._indexMap = new IndexMap({ tokenizer: this._tokenizer, fields: indexFields });
    this.indexKey = indexKey;
    this.indexFields = indexFields;
    this.caseSensitive = caseSensitive;
  }

  index(json: Record<string, unknown>): void {
    const key = json[this.indexKey];
    if (key && typeof key === 'string' && key.length > 0) {
      const mappedJson: Record<string, unknown> = {};
      for (const field of this.indexFields) {
        if (json.hasOwnProperty(field)) {
          mappedJson[field] = json[field];
        }
      }
      if (Object.keys(mappedJson).length > 0) {
        this._indexMap.add(key, mappedJson);
      }
    }
  }

  search(text: string): SearchResult[] {
    const words = text.split(this._tokenizer.delimiter);
    const result: DocPropIndex[] = [];
    for (const word of words) {
      const element = this.caseSensitive ? word : word.toLowerCase();
      const found = this._indexMap.find(element);
      if (found) {
        result.push(...found);
      }
    }
    return result.map((e) => new SearchResult(this._indexMap.keyOf(e[0]), this._indexMap.fieldOf(e[1])));
  }

  exportIndex(): string {
    return this._indexMap.export();
  }

  importFromJson(json: { keys: string[]; data: Record<string, DocPropIndex[]>; fields: string[] }): void {
    this._indexMap.import(json);
  }
}

export { SearchResult, TenoTextSearch };
